torchao/quantization/GPTQ.py

```diff
@@ -528,6 +528,9 @@ def _check_linear_int4_k(k, groupsize = 1, inner_k_tiles = None):
 def linear_forward_int4(x, weight_int4pack, scales_and_zeros, out_features, groupsize, dtype=torch.bfloat16):
     origin_x_size = x.size()
     x = x.reshape(-1, origin_x_size[-1])
+    if "cuda" in str(x.device):
+        # _weight_int4pack_mm CUDA kernel only supports bfloat16
+        dtype = torch.bfloat16
```

Comment:
this feels hacky, can you fix this properly?

Reply:
CUDA kernel only accepts bfloat16 AFAIK: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/cuda/int4mm.cu#L1144

> just setting dtype explicitly outside the function call?

I don't think that's a good experience, because it requires the user of this quantization API to know about the implementation detail of `_weight_int4pack_mm_cuda`.

Reply:
@jerryzh168 @malfet any suggestions?

Reply:
this means we should error out when `x` is not `bfloat16` right? I'm not sure why we need to implicitly convert this here

Reply:
this won't be an issue if we move to the tensor subclass solution probably, since we have fallback path for float32

Reply:
OK how about I left a TODO here and clean up in the future? AFAIK we are moving to new tensor subclass API and this code won't be used anyways.